File: api-js/src/SDKMeta.ts
import sdkLanguages from './data/languages.json'
import sdkRepos from './data/repos.json'
import sdkNames from './data/names.json'
import sdkTypes from './data/types.json'
import sdkPopularity from './data/popularity.json'
import sdkReleases from './data/releases.json'
import sdkUserAgents from './data/user_agents.json'
import sdkAiSdkIdentifiers from './data/ai_sdk_identifiers.json'

export enum Type {
    // ClientSide is an SDK that runs in a client scenario.
    ClientSide = "client-side",
    // ServerSide is an SDK that runs in a server scenario.
    ServerSide = "server-side",
    // Edge is an SDK that runs in an edge deployment scenario.
    Edge = "edge",
    // AI is an SDK that is primarily focused on AI/ML use cases.
    AI = "ai",
    // OpenFeatureProvider is an OpenFeature provider.
    OpenFeatureProvider = "open-feature-provider",
    // Relay is Relay Proxy.
    Relay = "relay",
    // Unknown if the SDK's type is not recognized.
    Unknown = "unknown"
}

function isType(value: string): value is Type {
    return Object.values(Type).includes(value as Type);
}

export type Repo = {
    github: string;
}

export const Languages: Record<string, string[]> = sdkLanguages;
export const Names: Record<string, string> = sdkNames;
export const Repos: Record<string, Repo> = sdkRepos;
export const Popularity: Record<string, number> = sdkPopularity;
export const Releases: ReleaseList = Object.fromEntries(
    Object.entries(sdkReleases).map(([key, value]) => [
      key,
      value.map((release: any) => ({
          Major: release["major"],
          Minor: release["minor"],
          Date: new Date(release["date"]),
          EOL: release["eol"] ? new Date(release["eol"]) : null
      }))
    ]));

export const Types: Record<string, Type> = Object.fromEntries(
    Object.entries(sdkTypes).map(([key, value]) => [
      key,
      isType(value) ? value : Type.Unknown
    ]));


export interface Release {
    Major: number;
    Minor: number;
    Date: Date;
    EOL: Date | null;
}

export interface ReleaseList {
    [key: string]: Release[];
}

export namespace ReleaseHelpers {
    export const IsLatest = (release: Release) => release.EOL === null;
    export const IsEOL = (release: Release, now: Date) => !IsLatest(release) && now > release.EOL!;
    export const IsApproachingEOL = (release: Release, now: Date, thresholdPrior: number) =>
        !IsLatest(release) && now.getTime() + thresholdPrior > release.EOL!.getTime();

    export const Earliest = (releases: Release[]) => releases[releases.length - 1];
    export const Latest = (releases: Release[]) => releases[0];
}

export interface UserAgent {
    userAgents?: string[];
    wrapperNames?: string[];
}

export const UserAgents: Record<string, UserAgent> = sdkUserAgents;

export namespace UserAgentHelpers {
    /**
     * Attempts to find an SDK name by checking wrapper names and user agents.
     * First checks wrapper names, then user agents, in alphabetical order by SDK ID.
     * 
     * @param identifier - The wrapper name or user agent string to search for
     * @returns The SDK name if found, undefined if not found
     */
    export const getSDKNameByWrapperOrUserAgent = (identifier: string): string | undefined => {
        const sdkId = getSDKIDByWrapperOrUserAgent(identifier);
        return sdkId === undefined ? undefined : Names[sdkId];
    }

    /**
     * Attempts to find an SDK ID by checking wrapper names and user agents.
     * First checks wrapper names, then user agents, in alphabetical order by SDK ID.
     *
     * Prefer this over getSDKNameByWrapperOrUserAgent when you need more than a display
     * name. The ID is the key for Names, Types, Repos, and Releases.
     *
     * Some identifiers are reported by more than one SDK. For example both akamai-base
     * and akamai-edgekv report "AkamaiEdgeSDK". In that case the first SDK ID in
     * alphabetical order is returned.
     *
     * @param identifier - The wrapper name or user agent string to search for
     * @returns The SDK ID if found, undefined if not found
     */
    export const getSDKIDByWrapperOrUserAgent = (identifier: string): string | undefined => {
        // Sort the entries by SDK ID to ensure consistent ordering.
        const sortedEntries = Object.entries(UserAgents).sort(([a], [b]) => a.localeCompare(b));

        // First check wrapper names
        for (const [sdkId, info] of sortedEntries) {
            if (info.wrapperNames?.includes(identifier)) {
                return sdkId;
            }
        }

        // Then check user agents
        for (const [sdkId, info] of sortedEntries) {
            if (info.userAgents?.includes(identifier)) {
                return sdkId;
            }
        }

        return undefined;
    }
}

/**
 * One (name, language) pair that an AI SDK reports about itself in the
 * $ld:ai:sdk:info custom event.
 */
export interface AISDKIdentifier {
    /** The value reported as aiSdkName. This is the package name. */
    name: string;
    /** The value reported as aiSdkLanguage, for example "python". */
    language: string;
}

/**
 * A map of SDK IDs to the identifiers those AI SDKs report.
 *
 * An AI SDK wraps a client that the caller supplies, so it sends no user agent of its
 * own. It reports itself in a custom event instead. Historical names are kept so that
 * older deployments still resolve.
 */
export const AISDKIdentifiers: Record<string, AISDKIdentifier[]> = sdkAiSdkIdentifiers;

export namespace AISDKHelpers {
    /**
     * Finds the SDK ID for an AI SDK from the aiSdkName and aiSdkLanguage it reports in
     * the $ld:ai:sdk:info event.
     *
     * Both arguments are required. The name alone is not unique: the Python and Ruby AI
     * SDKs both report "launchdarkly-server-sdk-ai".
     *
     * The returned ID is the key for Names, Types, Repos, and Releases.
     *
     * @param name - The value reported as aiSdkName
     * @param language - The value reported as aiSdkLanguage
     * @returns The SDK ID if found, undefined if not found
     */
    export const resolveAISDK = (name: string, language: string): string | undefined => {
        // Sort the entries by SDK ID to ensure consistent ordering.
        const sortedEntries = Object.entries(AISDKIdentifiers).sort(([a], [b]) => a.localeCompare(b));

        for (const [sdkId, identifiers] of sortedEntries) {
            if (identifiers.some((i) => i.name === name && i.language === language)) {
                return sdkId;
            }
        }

        return undefined;
    }
}

File: api/sdkmeta/sdkmeta.go
package sdkmeta

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"sort"
	"time"
)

//go:embed data/names.json
var namesJSON []byte

// Names is a map of SDK IDs to display names.
var Names map[string]string

//go:embed data/repos.json
var reposJSON []byte

// Repo contains the location of an SDK.
type Repo struct {
	// GitHub is the GitHub repo path including the owner and repo name (e.g. launchdarkly/js-core).
	GitHub string `json:"github"`
}

// Repos is a map of SDK IDs to repository information.
var Repos map[string]Repo

//go:embed data/languages.json
var languagesJSON []byte

// Languages is a map of SDK IDs to supported languages.
var Languages map[string][]string

//go:embed data/popularity.json
var popularityJSON []byte

// Popularity is a map of SDK IDs to popularity scores.
var Popularity map[string]int

//go:embed data/types.json
var typesJSON []byte

// Type represents the most common use-case for an SDK.
type Type string

const (
	// ClientSideType is an SDK that runs in a client context.
	ClientSideType Type = "client-side"
	// ServerSideType is an SDK that runs in a server context.
	ServerSideType Type = "server-side"
	// EdgeType is an SDK that runs in an edge deployment scenario.
	EdgeType Type = "edge"
	// AIType is an SDK that is primarily focused on AI/ML use cases.
	AIType Type = "ai"
	// OpenFeatureProviderType is an OpenFeature provider.
	OpenFeatureProviderType Type = "open-feature-provider"
	// RelayType is Relay Proxy.
	RelayType Type = "relay"
)

// Types is a map of SDK IDs to SDK types.
var Types map[string]Type

//go:embed data/releases.json
var releasesJSON []byte

type Release struct {
	Major int        `json:"major"`
	Minor int        `json:"minor"`
	Date  time.Time  `json:"date"`
	EOL   *time.Time `json:"eol"`
}

// MajorMinor returns a version string comprised of the major and minor version. For example,
// '2.1'.
func (r Release) MajorMinor() string {
	return fmt.Sprintf("%d.%d", r.Major, r.Minor)
}

// IsLatest returns true if the release is the latest release, meaning there is no EOL date set.
func (r Release) IsLatest() bool {
	return r.EOL == nil
}

// IsEOL returns true if the release is not the latest release and the current time is after the EOL date. The parameter
// represents the current time.
func (r Release) IsEOL(now time.Time) bool {
	return !r.IsLatest() && now.After(*r.EOL)
}

// IsApproachingEOL returns true if the release is not the latest release and the EOL date is within the time period
// from now to now + thresholdPrior. This is only valid if IsEOL() returns false.
func (r Release) IsApproachingEOL(now time.Time, thresholdPrior time.Duration) bool {
	return !r.IsLatest() && now.Add(thresholdPrior).After(*r.EOL)
}

// ReleaseList is an ordered list of releases. The first item should be the most recent release, while the
// last item is the oldest release.
type ReleaseList []Release

var Releases map[string]ReleaseList

// Earliest returns the earliest release.
func (r ReleaseList) Earliest() Release {
	return r[len(r)-1]
}

// Latest returns the latest release.
func (r ReleaseList) Latest() Release {
	return r[0]
}

//go:embed data/user_agents.json
var userAgentsJSON []byte

// SDKUserAgentMap contains user agent and wrapper information for an SDK
type SDKUserAgentMap struct {
	UserAgents   []string `json:"userAgents,omitempty"`
	WrapperNames []string `json:"wrapperNames,omitempty"`
}

// UserAgentMap maps SDK IDs to the user agents and wrapper names they report.
type UserAgentMap map[string]SDKUserAgentMap

// UserAgents is a map of SDK IDs to their user agent and wrapper information
var UserAgents UserAgentMap

// ResolveSDKID finds the SDK ID that reports the given wrapper name or user agent.
// Wrapper names are checked first, then user agents, in alphabetical order by SDK ID.
// Returns the SDK ID and true if found, empty string and false if not found.
//
// The returned ID is the key for Names, Types, Repos, and Releases.
//
// Some identifiers are reported by more than one SDK. For example both akamai-base and
// akamai-edgekv report "AkamaiEdgeSDK". In that case the first SDK ID in alphabetical
// order is returned.
func (m UserAgentMap) ResolveSDKID(identifier string) (string, bool) {
	// Get sorted SDK IDs to ensure consistent ordering
	var sdkIDs []string
	for sdkID := range m {
		sdkIDs = append(sdkIDs, sdkID)
	}
	sort.Strings(sdkIDs)

	// First check wrapper names
	for _, sdkID := range sdkIDs {
		for _, wrapper := range m[sdkID].WrapperNames {
			if wrapper == identifier {
				return sdkID, true
			}
		}
	}

	// Then check user agents
	for _, sdkID := range sdkIDs {
		for _, agent := range m[sdkID].UserAgents {
			if agent == identifier {
				return sdkID, true
			}
		}
	}

	return "", false
}

// GetSDKNameByWrapperOrUserAgent attempts to find an SDK name by first checking wrapper names,
// then user agents, in alphabetical order by SDK ID. Returns the SDK name and true if found,
// empty string and false if not found.
//
// Use UserAgents.ResolveSDKID when you need more than a display name.
func GetSDKNameByWrapperOrUserAgent(identifier string) (string, bool) {
	sdkID, ok := UserAgents.ResolveSDKID(identifier)
	if !ok {
		return "", false
	}
	return Names[sdkID], true
}

//go:embed data/ai_sdk_identifiers.json
var aiSdkIdentifiersJSON []byte

// AISDKIdentifier is one (name, language) pair that an AI SDK reports about itself
// in the $ld:ai:sdk:info custom event.
type AISDKIdentifier struct {
	// Name is the value reported as aiSdkName. This is the package name.
	Name string `json:"name"`
	// Language is the value reported as aiSdkLanguage, for example "python".
	Language string `json:"language"`
}

// AISDKIdentifierMap maps SDK IDs to the identifiers those AI SDKs report.
type AISDKIdentifierMap map[string][]AISDKIdentifier

// AISDKIdentifiers is a map of SDK IDs to the identifiers those AI SDKs report.
//
// An AI SDK wraps a client that the caller supplies, so it sends no user agent of its
// own. It reports itself in a custom event instead. Historical names are kept so that
// older deployments still resolve.
var AISDKIdentifiers AISDKIdentifierMap

// ResolveSDKID finds the SDK ID for an AI SDK from the aiSdkName and aiSdkLanguage it
// reports in the $ld:ai:sdk:info event. Returns the SDK ID and true if found, empty
// string and false if not found.
//
// Both arguments are required. The name alone is not unique: the Python and Ruby AI
// SDKs both report "launchdarkly-server-sdk-ai".
//
// The returned ID is the key for Names, Types, Repos, and Releases.
func (m AISDKIdentifierMap) ResolveSDKID(name string, language string) (string, bool) {
	// Get sorted SDK IDs to ensure consistent ordering
	var sdkIDs []string
	for sdkID := range m {
		sdkIDs = append(sdkIDs, sdkID)
	}
	sort.Strings(sdkIDs)

	for _, sdkID := range sdkIDs {
		for _, identifier := range m[sdkID] {
			if identifier.Name == name && identifier.Language == language {
				return sdkID, true
			}
		}
	}

	return "", false
}

func panicOnError(err error) {
	if err != nil {
		panic("couldn't initialize SDK Metadata module: " + err.Error())
	}
}

func init() {
	panicOnError(json.Unmarshal(namesJSON, &Names))
	panicOnError(json.Unmarshal(reposJSON, &Repos))
	panicOnError(json.Unmarshal(languagesJSON, &Languages))
	panicOnError(json.Unmarshal(typesJSON, &Types))
	panicOnError(json.Unmarshal(releasesJSON, &Releases))
	panicOnError(json.Unmarshal(popularityJSON, &Popularity))
	panicOnError(json.Unmarshal(userAgentsJSON, &UserAgents))
	panicOnError(json.Unmarshal(aiSdkIdentifiersJSON, &AISDKIdentifiers))
}
